add group_inter_qubit_commuting

YumaNK opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This pull request is addressed for #6690. 
The new feature for measurement operator commutation is based on #7651 

### Details and comments

Mentored by @ikkoham

HuangJunye commented:
@YumaNK Thanks for making this contribution. It's really impressive to make a PR so early in the mentorship program. Linking the QAMP issue here to keep track. https://github.com/qiskit-advocate/qamp-spring-22/issues/20

YumaNK commented on qiskit/quantum_info/operators/symplectic/pauli_list.py:
```diff
@@ -1085,9 +1088,15 @@ def _noncommutation_graph(self):
         )
         mat2 = mat1[:, None]
         # mat3[i, j] is True if i and j are qubit-wise commutable
-        mat3 = (((mat1 * mat2) * (mat1 - mat2)) == 0).all(axis=2)
+        mat3 = (mat1 * mat2) * (mat1 - mat2)
```
I see. I will make unit tests for the functions I changed.

ikkoham commented:
I discussed this PR with @YumaNK. I proposed to deprecate `group_qubitwise_commuting` in `PauliList` and have the general method `group_commuting(self, qubit_wise=True)`.

ikkoham commented:
I discussed this PR with @YumaNK. I proposed to deprecate `group_qubitwise_commuting` in `PauliList` and have the general method `group_commuting(self, qubit_wise=True)`.

jakelishman commented:
I'm not sure I see a reason to deprecate the other method - we only added it recently, and it isn't causing any problems. Changing the way users access something should be avoided unless necessary, for the most part, because it just creates friction for them.

At the very least, we can't deprecate the other method until at least Terra 0.22, assuming this method (with the new keyword argument) gets released in Terra 0.21. I personally wouldn't bother deprecating or adding the keyword arg, but I don't feel strongly about it. There's no real issue with having two public functions that call the same private worker routine.

YumaNK commented:
For the use case, we expect to reduce the burden of VQE calculation. As the Hamiltonian is decomposed into Pauli operators, grouping Pauli operator enables to find the fewest iterations for evaluating expectation value through measurement.

In Qiskit Tutorial for operator_flow, AbelianGrouper is mentioned for calculation Hamiltonian of H2 molecule in VQE. 
https://qiskit.org/documentation/stable/0.26/tutorials/operators/01_operator_flow.html#Expectations
We are aiming to implement such a function in PauliList and SparsePauliOp classes in an even more elaborated and faster way.

For the discussion about deprecation, we decide to add group_commuting method in PauliList as well as group_qubitwise_commuting. Instead of deprecating group_qubitwise_commuting, we will wrap it by group_commuting, a more general method,  with option qubit_wise=True.

jakelishman commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -772,6 +774,58 @@ def __getitem__(self, key):
+
+                For example:
+
+                    * ``SparsePauliOp.from_list([('XX', 2),``
+                    * ``                               ('YY', 1),``
+                    * ``                               ("IZ",2j),``
+                    * ``                               ("ZZ",1j)]).group_commuting()``
+                            Returns [SparsePauliOp(['IZ', 'ZZ'], coeffs=[0.+2.j, 0.+1.j]),
+                                            SparsePauliOp(['XX', 'YY'], coeffs=[2.+0.j, 1.+0.j])]
```
This won't render correctly in the documentation.  You can build the documentation with `tox -edocs` to test.  I imagine you mean something like
```text
Args:
    qubit_wise (bool): whether the commutation rule is applied to the whole operator,
        or on a per-qubit basis.  For example:

        .. code-block:: python

            >>> op = SparsePauliOp.from_list([("XX", 2), ("YY", 1), ...])
            >>> op.group_commuting()
            [SparsePauliOp(["IZ", "ZZ"], coeffs=[0.+2.j, 0.+1j]),
             SparsePauliOp(["XX", "YY"], coeffs=[2.+0.j, 1.+0.j])]
            >>> op.group_commuting(qubit_wise=True)
            [...]
```

YumaNK replied:
Thank you for instructing documentation syntax. This expression and test command are exactly what I sought for.
So far, I will adopt your suggested wording of `qubit_wise` argument and be consistent with that wording of dosctrings over methods which use the argument.

jakelishman commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -772,6 +774,58 @@ def __getitem__(self, key):
+        for idx, color in coloring_dict.items():
+            groups[color].append(idx)
+        return [self[group] for group in groups.values()]
+
+
+# Update docstrings for API docs
+generate_apidocs(SparsePauliOp)
 
 # Update docstrings for API docs
 generate_apidocs(SparsePauliOp)
```
Be careful with copy-pasting code here.

YumaNK replied:
I'm very sorry for that kinds of mistake..

jakelishman commented on test/python/quantum_info/operators/symplectic/test_pauli_list.py:
```diff
@@ -2058,20 +2058,55 @@ def qubitwise_commutes(left: Pauli, right: Pauli) -> bool:
         # checking that every input Pauli in pauli_list is in a group in the ouput
         output_labels = [pauli.to_label() for group in groups for pauli in group]
-        assert sorted(output_labels) == sorted(input_labels)
+        #     assert sorted(output_labels) == sorted(input_labels)
```
This assertion needs returning to the test.

YumaNK replied:
I rewrote that test case into 
`self.assertListEqual(sorted(output_labels), sorted(input_labels))`
and left old code as comment-out by my carelessness.
I will delete this line of comment.

jakelishman commented on test/python/quantum_info/operators/symplectic/test_pauli_list.py:
```diff
@@ -2058,20 +2058,55 @@ def qubitwise_commutes(left: Pauli, right: Pauli) -> bool:
         # Within each group, every operator qubit-wise commutes with every other operator.
         for group in groups:
-            assert all(
-                qubitwise_commutes(pauli1, pauli2)
-                for pauli1, pauli2 in itertools.combinations(group, 2)
+            self.assertTrue(
+                all(commutes(pauli1, pauli2) for pauli1, pauli2 in itertools.combinations(group, 2))
             )
```
Please be careful with the comments - it's vitally important that they are accurate to the code to help future maintainers understand what's going on, especially in tests and mathematical code.  Copy-pasting code is generally not a good idea, because it leads to mistakes like this.

There are a few other places in this PR at the moment where there are copy-pasted comments that aren't accurate to the surrounding code.  Please could you check through all the comments carefully, and make sure that they're accurate?  It might help to use the PR review panel on GitHub to show all the changes you've made, so you can check that the PR is in the form you intend it to be in.

jakelishman reviewed (commented):
Thanks for the explanation for the motivation of the PR.  I'm happy to accept it if Nature has a valid use-case in mind already.  The new API looks fine to me - things that are purely additive are usually best at first.  If it later becomes a problem that we have multiple ways of saying `group_commuting(qubit_wise=True)`, we can revisit a potential deprecation in the future, but that can't be for _at least_ another minor version of Terra, and I don't see any particular need for it myself.

The code changes in general look good here, but please have a look through and make sure that all the comments and documentation render correctly and are up-to-date.  That's just as important as the code working, because it'll be very hard to work out what's going on in the future if the comments aren't correct.

ikkoham commented:
This is a more general discussion, but we don't use `PendingDeprecationWarning` much in qiskit-terra. It's not even mentioned in the deprecation in the Contribution Guidelines. What about giving a `PendingDeprecationWarning` once?

jakelishman commented:
It's totally ok to use `PendingDeprecationWarning`, it just doesn't count towards the deprecation timer because it's not generally a visible warning.

My concern about deprecation isn't that we're not giving enough warning, it's that there doesn't seem to be any technical need to deprecate a feature we only just added. Every time we do a deprecation and removal, it annoys downstream users.  In this case, as long as one of the two methods just directly calls the other, I don't think there's any maintenance burden, and it lets us keep users happy.

YumaNK commented on qiskit/quantum_info/operators/symplectic/pauli_list.py:
```diff
@@ -1084,25 +1087,58 @@ def _noncommutation_graph(self):
-        mat3 = (((mat1 * mat2) * (mat1 - mat2)) == 0).all(axis=2)
-        # convert into list where tuple elements are qubit-wise non-commuting operators
-        return list(zip(*np.where(np.triu(np.logical_not(mat3), k=1))))
+        qubit_commutation_mat = (mat1 * mat2) * (mat1 - mat2)
+        # adjacency_mat[i, j] is True if i and j are commutable
+        if qubit_wise:
+            adjacency_mat = np.logical_or.reduce(qubit_commutation_mat, axis=2)
+        else:
+            adjacency_mat = np.logical_xor.reduce(qubit_commutation_mat, axis=2)
```
Thank you for careful review. I will change comment as following.
        # adjacency_mat[i, j] is True if an edge (i, j) is present, i.e. i and j are non-commuting

This description would fit with the docstring of its method ` _noncommutation_graph(self, qubit_wise)`

YumaNK commented:
@jakelishman 
I fixed inaccurate comments that accidentally inherited `group_inter_qubit_commuting`.
The release note was also added. Do I have things to check before merge?

jakelishman reviewed (approved):
Sorry for the delay here.  This looks good now, thanks for the updates!

jakelishman commented:
@YumaNK: I thought you might like to know that this PR caused a 10% performance improvement in our benchmarking - we didn't initially think to check that while reviewing, but your reorganisation of the helper function into better Numpy code seems to have paid off in both speed and features. Thanks! 🚀
